fleetspeak/src/admin/history/history.go

```diff
@@ -0,0 +1,202 @@
+	if err := validate(cs); err != nil {
+		return nil, err
+	}
+	if len(cs) == 0 {
+		return &Summary{}, nil
+	}
+
+	sort.Sort(contactSlice(cs))
+
+	nm := nonceMap(cs)
+	if err := validateTime(nm); err != nil {
+		return nil, err
+	}
+
+	splits, splitPoints := countSplits(nm)
+	skips := countSkips(cs)
+
+	start, _ := ptypes.Timestamp(cs[0].Timestamp)
+	end, _ := ptypes.Timestamp(cs[len(cs)-1].Timestamp)
+
+	return &Summary{
+		Start:       start,
+		End:         end,
+		Count:       len(cs),
+		IPCount:     countIPs(cs),
+		Splits:      splits,
+		SplitPoints: splitPoints,
+		Skips:       skips,
+	}, nil
+}
+
+func validate(contacts []*spb.ClientContact) error {
+	times := make(map[int64]bool)
+	nonces := make(map[uint64]bool)
+	for _, c := range contacts {
+		// The datastore assigns and is responsible for preventing duplicate contact
+		// timestamps. If we see a duplicate, it is a data error that might confuse
+		// other analysis.
+		t, err := ptypes.Timestamp(c.Timestamp)
+		if err != nil {
+			return fmt.Errorf("bad timestamp proto [%v]: %v", c.Timestamp, err)
+		}
+		ts := t.UnixNano()
+		if times[ts] {
+			return fmt.Errorf("duplicate timestamp: %v", t)
+		}
+		times[ts] = true
+
+		// The sent nonce is a 64 bit number that should have been choosen by the
+		// server for each contact using a strong RNG. If we see a duplicate it is a
+		// data error that might confuse other analysis.
+		if nonces[c.SentNonce] {
+			return fmt.Errorf("duplicate sent nonce: %d", c.SentNonce)
+		}
+		nonces[c.SentNonce] = true
+	}
+	return nil
+}
+
+func validateTime(nm map[uint64]*spb.ClientContact) error {
+
+	for _, i := range nm {
+		if i.ReceivedNonce == 0 {
+			continue
+		}
+		if p := nm[i.ReceivedNonce]; p != nil {
+			if !tsLess(p.Timestamp, i.Timestamp) {
+				// The nonce recieved from a client cannot reference a nonce produced by
+				// the server in the future. If this seems to have happened, it is a
+				// data error that could confuse other analysis.
+				return fmt.Errorf("Nonce at [%v] references future time [%v].", i.Timestamp, p.Timestamp)
+			}
+		}
+	}
+	return nil
+}
+
+func countIPs(contacts []*spb.ClientContact) int {
+	m := make(map[string]bool)
+	for _, c := range contacts {
+		h, _, err := net.SplitHostPort(c.ObservedAddress)
+		if err != nil {
+			log.Warningf("Unable to parse ObservedAddress [%s], ignoring: %v", c.ObservedAddress, err)
+			continue
+		}
+		m[h] = true
+	}
+	return len(m)
+}
+
+func nonceMap(cs []*spb.ClientContact) map[uint64]*spb.ClientContact {
+	n := make(map[uint64]*spb.ClientContact)
+	for _, c := range cs {
+		n[c.SentNonce] = c
+	}
+	return n
+}
+
+func countSplits(nm map[uint64]*spb.ClientContact) (splits, splitPoints int) {
+	// A count of how many contacts target each nonce.
+	ts := make(map[uint64]int)
+	for _, c := range nm {
+		// Anything not in nm must be too old - coalesce with the 0 target.
+		if c.ReceivedNonce == 0 || nm[c.ReceivedNonce] == nil {
+			ts[0] = ts[0] + 1
```

Comment:
Nit: why not += 1 or even ++ here and below?

Reply:
Done. (I wasn't 100% sure it'd work, but it seems fine.)